crates/amaru/src/stages/consensus/forward_chain/tcp_forward_chain_server.rs

```diff
@@ -0,0 +1,211 @@
+    #[expect(clippy::expect_used)]
+    #[expect(clippy::unwrap_used)]
+    pub fn new(
+        store: Arc<dyn ChainStore<Header>>,
+        listen_address: String,
+        network_magic: u64,
+        max_peers: usize,
+        our_tip: HeaderTip,
+    ) -> anyhow::Result<Self> {
+        let runtime =
+            AcTokio::new("consensus.forward", 1).expect("failed to create AcTokio runtime");
+
+        // Start a TCP listener synchronously but in non-blocking mode
+        // so we can integrate it into the Tokio runtime later.
+        // This allows us to create the TcpForwardChainServer in a synchronous context.
+        let std_listener = std::net::TcpListener::bind(&listen_address)?;
+        std_listener.set_nonblocking(true)?;
+
+        let tcp_listener = runtime
+            .with_rt(|rt| {
+                let _e = rt.enter(); // attach to that runtime
+                tokio::net::TcpListener::from_std(std_listener)
+                    .expect("failed to wrap std listener")
+            })
+            .unwrap();
+
+        TcpForwardChainServer::create(
+            store,
+            Arc::new(runtime),
+            tcp_listener,
+            network_magic,
+            max_peers,
+            our_tip,
+        )
+    }
+
+    /// Creates a new TcpForwardChainServer instance with a provided Acto runtime and TcpListener.
+    #[expect(clippy::unwrap_used)]
+    pub fn create(
+        store: Arc<dyn ChainStore<Header>>,
+        runtime: Arc<AcTokio>,
+        tcp_listener: TcpListener,
+        network_magic: u64,
+        max_peers: usize,
+        our_tip: HeaderTip,
+    ) -> anyhow::Result<Self> {
+        let clients = runtime
+            // FIXME: This is a temporary stop gap solution while we wait
+            // to refactor to use pure-stage. Acto library has as a capped
+            // size for the mailbox and drops incoming messages when it's
+            // full.  This should not be a problem in real life, but while
+            // we are syncing _and_ forwarding at the same time for demo
+            // purpose, this is problematic.
+            .with_mailbox_size(1_000_000)
+            .spawn_actor("chain_forward", |cell| {
+                client_supervisor(cell, store.clone(), max_peers)
+            })
+            .me;
+
+        let our_tip = Arc::new(Mutex::new(our_tip.clone()));
+        runtime.with_rt(|rt| {
+            let our_tip_clone = our_tip.clone();
+            let clients_clone = clients.clone();
+            rt.spawn(async move {
+                loop {
+                    // due to the signature of TcpListener::accept, this is the only way to use this API
+                    // in particular, it isn’t possible to poll for new peers within the `schedule` method
+                    match PeerServer::accept(&tcp_listener, network_magic).await {
+                        Ok(peer) => {
+                            let our_tip = our_tip_clone.lock().unwrap().clone();
```

Comment:
This seems off: why would we inject into the connection notification the current state of the actor (or something close to it), when the client will tell us later where it is at? Let the information flow naturally to the one place where the pieces can be combined and acted upon.

Reply:
Isn't that the place where we communicate our tip to the client upon connection? Are you saying that this is not necessary?